tests/lib/tools/mkimage-uc22

```diff
@@ -0,0 +1,566 @@
+    printf "%s" "$GRUBENV" > "$CACHE"/grubenv
+    sudo cp -a "$CACHE"/grubenv "$MNT"/ubuntu-boot/EFI/ubuntu/grubenv
+    sudo mkdir -p "$MNT"/ubuntu-boot/device/
+    sudo cp -a "$MODEL_ASSERTION" "$MNT"/ubuntu-boot/device/model
+
+    # kernel
+    sudo mkdir -p "$MNT"/ubuntu-boot/EFI/ubuntu/"$KERNEL_SNAP"
+    sudo cp -a "$CACHE"/snap-pc-kernel/kernel.efi "$MNT"/ubuntu-boot/EFI/ubuntu/"$KERNEL_SNAP"
+    sudo ln -sf "$KERNEL_SNAP"/kernel.efi "$MNT"/ubuntu-boot/EFI/ubuntu/kernel.efi
+
+    # cleanup
+    sync
+    sudo umount "$MNT"/ubuntu-*
+}
+
+replace_initramfs_bits() {
+    CACHE="$1"
+    SNAPD_BINPATH="$2"
+    # TODO is there code to share with uc20_build_initramfs_kernel_snap in prepare.sh?
+    KERNEL_EFI_ORIG="$CACHE"/snap-pc-kernel/kernel.efi
+    if [ ! -d initrd ]; then
+        objcopy -O binary -j .initrd "$KERNEL_EFI_ORIG" initrd.img
+        unmkinitramfs initrd.img initrd
+    fi
+
+    # Retrieve efi stub from ppa so we can rebuild kernel.efi
+    sudo DEBIAN_FRONTEND=noninteractive apt install -y --no-install-recommends ubuntu-dev-tools lz4 < /dev/null > /dev/null
+    pull-lp-debs -a amd64 -D ppa --ppa ppa:snappy-dev/image ubuntu-core-initramfs "jammy"
+    dpkg --fsys-tarfile ubuntu-core-initramfs_*.deb |
+        tar --wildcards -xf - './usr/lib/ubuntu-core-initramfs/efi/linux*.efi.stub'
+
+    cp "$SNAPD_BINPATH"/snap-bootstrap initrd/main/usr/lib/snapd/
+    ( cd initrd/main && find . | cpio --create --quiet --format=newc --owner=0:0 | lz4 -l -7 > ../../initrd.img.new )
+
+    objcopy -O binary -j .linux "$KERNEL_EFI_ORIG" linux
+    # Replace kernel.efi in unsquashed snap
+    objcopy --add-section .linux=linux --change-section-vma .linux=0x2000000 \
+            --add-section .initrd=initrd.img.new --change-section-vma .initrd=0x3000000 \
+            usr/lib/ubuntu-core-initramfs/efi/linux*.efi.stub \
+            "$KERNEL_EFI_ORIG"
+}
+
+cleanup() {
+    IMG="$(readlink -f "$1")"
+    MNT="$(readlink -f "$2")"
+
+    sync
+    sleep 1
+    umount "$MNT"/* || true
+    sleep 1
+    kpartx -d "$IMG" || true
```

Comment:
Maybe use `retry` here ?

Reply:
I updated the cleanup function but for some raeson it was repeated, so I deleted this one. Thanks for the comment